lom/src/engine/sequenceHandler.go

```diff
@@ -0,0 +1,832 @@
+        seq.sequenceStatus = sequenceStatus_complete
+        return
+    }
+
+    errCode, errStr, err = p.resumeSequence(seq)
+    if errCode == LoMActionActive {
+        /*
+         * May be a prior seq left behind an active request. We got to wait
+         * till it responds or this sequence timeout, whichever earlier.
+         */
+         errCode = LoMResponseOk
+     }
+    return
+}
+
+/*
+ * resumeSequence
+ *
+ * Validate ctIndex. Attempt to raise request per ctIndex. If failed, return
+ * appropriate error. If succeeded, add it to corresponding client's pending
+ * server requests and list of active requests. Set it as current request of
+ * this sequence.
+ */
+func (p *SeqHandler_t) resumeSequence(seq *sequenceState_t) (errCode LoMResponseCode, errStr string, err error) {
+    regF := GetRegistrations()
+
+    errCode = LoMResponseOk
+    errStr = ""
+    err = nil
+
+    if seq == nil {
+        errCode = LoMSequenceIncorrect
+        errStr = "Resume called on nil sequence"
+        return
+    }
+    if !seq.validate() {
+        errCode = LoMSequenceIncorrect
+        errStr = fmt.Sprintf("Resume called on invalid seq (%v)", *seq)
+        return
+    }
+    if seq.currentRequest != nil {
+        LogInfo("current req is active. Nothing to resume (%v)", *seq)
+        return
+    }
+
+    nextAction := seq.sequence.Actions[seq.ctIndex]
+
+    /* validate action */
+    if regF.GetActiveActionInfo(nextAction.Name) == nil {
+        errCode = LoMActionNotRegistered
+        errStr = fmt.Sprintf("%v", LogError("%s: %s not registered. Abort", seq.sequence.SequenceName,
+                    nextAction.Name))
+        return
+    } else if _, ok := p.activeRequests[nextAction.Name]; ok {
+        errCode = LoMActionActive
+        errStr = fmt.Sprintf("%v", LogError("%s: %s (%s). Abort", seq.sequence.SequenceName,
+                nextAction.Name, GetLoMResponseStr(LoMActionActive)))
+        return
+    }
+
+
+    /* All clear. Fire request for the next action in sequence */
+    req := &ServerRequestData {
+        TypeServerRequestAction,
+        &ActionRequestData  {
+            Action: nextAction.Name,
+            InstanceId: GetUUID(),
+            AnomalyInstanceId: seq.anomalyInstanceId,
+            AnomalyKey: seq.context[0].AnomalyKey,
+            Timeout : nextAction.Timeout,
+            Context: seq.context,
+        },
+    }
+
+    /* Add to client's pending Q  to send to client upon client asking for a request. */
+    /* Stays in this Q until client reads it */
+    if err := regF.AddServerRequest(nextAction.Name, req); err != nil {
+        errCode = LoMInternalError
+        errStr = fmt.Sprintf("%v", err)
+    } else {
+        /* Track it in our active requests;  Waits here till response */
+        tout := time.Now().Unix() + int64(nextAction.Timeout)
+        act := &activeRequest_t {req, seq.anomalyInstanceId, tout}
+        p.activeRequests[nextAction.Name] = act
+
```

Comment:
should we also set p.currentSequence to seq ?

Reply:
This is just resume a sequence, called from few places. When called from resumeNextSequence, it sets it as current. In other cases, it just called to invoke next action in this seq.

Reply:
but I think resume method should take care of setting p.currentSequence instead of caller so that consistency would be maintained for each call.

Reply:
Resume sequence is to activate next action on the sequence.
ResumeNextSequence is the one that kicks off a new sequence and hence the *only* authority to set the current